server/dynamo.go

```diff
@@ -108,3 +124,64 @@ func (c *Server) redeemTokenWithDynamo(issuer *Issuer, preimage *crypto.TokenPre
 	}
 	return nil
 }
+
+// PersistRedemption saves the redemption in the database
+func (c *Server) PersistRedemption(redemption RedemptionV2) error {
+	av, err := dynamodbattribute.MarshalMap(redemption)
+	if err != nil {
+		c.Logger.Error("Error marshalling redemption")
+		return err
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:                av,
+		ConditionExpression: aws.String("attribute_not_exists(id)"),
+		TableName:           aws.String("redemptions"),
+	}
+
+	_, err = c.dynamo.PutItem(input)
+	if err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "ConditionalCheckFailedException" { // unique constraint violation
+			c.Logger.Error("Duplicate redemption")
+			return errDuplicateRedemption
+		}
+		c.Logger.Error("Error creating item")
+		return err
+	}
+	return nil
+}
+
+// CheckRedeemedTokenEquivalence returns whether just the ID of a given RedemptionV2 token
+// matches an existing persisted record, the whole value matches, or neither match and
+// this is a new token to be redeemed.
+func (c *Server) CheckRedeemedTokenEquivalence(issuer *Issuer, preimage *crypto.TokenPreimage, payload string) (*RedemptionV2, Equivalence, error) {
+	preimageTxt, err := preimage.MarshalText()
+	if err != nil {
+		c.Logger.Error("Error Marshalling preimage")
+		return nil, UnknownEquivalence, err
+	}
+
+	id := uuid.NewSHA1(*issuer.ID, preimageTxt)
+
+	redemption := RedemptionV2{
+		IssuerID:  issuer.ID.String(),
+		ID:        id.String(),
+		PreImage:  string(preimageTxt),
+		Payload:   payload,
+		Timestamp: time.Now(),
```

Comment:
we will need to remove this, currently we will never get equivalence

Reply:
( or simply skip checking for timestamp equivalence )

Reply:
I've made the equivalence check omit the timestamp from comparison.